File: opcs-web/src/app/controller/security/role.controller.ts
import {RoleService} from "../../service/security/role/role.service";
import {map, Observable} from "rxjs";
import {Injectable} from "@angular/core";
import {RoleModel} from "../../model/security/role.model";
import {PaginationRequestModel} from "../../model/pagination/pagination.request.model";
import {PaginationResponseModel} from "../../model/pagination/pagination.response.model";
import {QueryRoleSearch} from "../../model/pagination/search/security/query-role.search";

@Injectable({
    providedIn: 'root'
})
export class RoleController {
    constructor(
        private readonly roleService: RoleService
    ) {}

	queryPage(pageModel: PaginationRequestModel<QueryRoleSearch>): Observable<PaginationResponseModel<RoleModel>> {
		return this.roleService.queryRole("role/queryPage", pageModel).pipe(
			map(response => response.result )
		)
	}

    getAllRole(): Observable<Array<RoleModel>> {
        return this.roleService.getAllRole("role/getAll").pipe(
            map((response) => response.result)
        )
    }
}

File: opcs/App/Controller/Security/RoleController.cs
using System.ComponentModel.DataAnnotations;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using opcs.App.Core.Attribute;
using opcs.App.Core.Security;
using opcs.App.Data;
using opcs.App.Data.Dto.Pagination;
using opcs.App.Data.Dto.Pagination.Search;
using opcs.App.Service.User.Interface;
using opcs.Resources;

namespace opcs.App.Controller.Security;

[Route("role")]
[Authorize]
[ApiController]
[Produces("application/json"), RequiredPermission(Permission.PermissionEnum.ManageRole)]
public class RoleController(
    IRoleService roleService,
    ILogger<RoleController> logger) : ControllerBase
{
    [HttpGet("getAll")]
    public IActionResult GetAllRoles()
    {
        return new Response { dto = roleService.GetAllRole() };
    }

    [HttpGet]
    public IActionResult GetRoleById([FromQuery, Required] int id)
    {
        return roleService.GetRoleById(id).Match(
            result => new Response { dto = result },
            () => new Response
            {
                statusCode = StatusCodes.Status400BadRequest,
                message = CodeMessages.opcs_error_role_not_exist
            });
    }

    [HttpPost("queryPage")]
    public IActionResult QueryRole(
        [FromBody, AssertSortField<QueryRoleSearch>(QuerySortField.SortFieldEnum.RoleSortField)]
        PaginationRequestDto<QueryRoleSearch> requestDto)
    {
        return new Response
        {
            dto = roleService.QueryRole(requestDto)
        };
    }
}